### DISCUSSION TITLE
Support for macros in Lua?

### DISCUSSION THREAD

**Discussion by pdenapo** [tectonic-typesetting/tectonic#1159] [General]
I see that tectonic is based on XeTeX. But there is another TeX engine out there, called Luatex 
that supports using macros in Lua.

This is great because TeX is not very nice as a programming language. And Lua is an scripting  
language which is popular and has and designed to be embedded in other applications.

See for instance this nice example of how to generate the Pascal triangle using tiks
(and compare it with the pure latex version of it)

https://tex.stackexchange.com/questions/17522/pascals-triangle-in-tikz

I wanted to ask if tectonic plans to support something like this. It would be nice to do it in Luatex
compatible way.

**Comment by chupinmaxime** [comment]
I do agree. Tectonic seems to be very usefull but LuaTeX is used a lot. Unfortunately, most of my LaTeX file cannot be compiled with XeLaTeX. 
Maybe tectonic could be built on LuaTeX?

**Comment by dbitouze** [comment]
I fully agree: LuaTeX is more and more widely used and is well maintained (by contrast to XeTeX, AFAIK).

**Comment by CraftSpider** [comment]
There's slow effort to move Tectonic to using basically its own custom backend engine. If pkgw is supportive, once that's done I'd be willing to look into supporting at least some of LuaTeX's primitives. However it's unlikely for Tectonic to port to LuaTeX directly, as that would be as much or more work for a lateral move.

Even if some support was added, it would involve a good bit of discussion - shipping a Lua interpreter with Tectonic would be a good chunk of code, and might not be something included in default builds.

**Comment by kpym** [comment]
While LuaTeX certainly has its advantages, especially with Lua scripting capabilities, my personal preference is to stick with XeTeX, primarily due to compilation speed. When comparing the compilation times across engines, we observe the following hierarchy:

`tectonic < pdflatex < xelatex < lualatex`

If we require the same level of PDF compression, XeLaTeX becomes nearly as fast as PDFLaTeX, while LuaLaTeX slows down even further. This performance difference is crucial for workflows that involve large or complex documents.

Given this, I am not in favor of a full switch to LuaTeX in Tectonic. However, exploring selective support for some LuaTeX primitives, as mentioned, could be a worthwhile compromise—so long as it does not negatively impact compilation time.

**Reply by dbitouze** [reply]
I understand, but Lua(La)TeX is not as slow as it used to be. Also, there are more and more Lua(La)TeX-specific packages, and as a result more and more users are switching to Lua(La)TeX exclusively. So `Tectonic` and all the tools derived from it (such as `TeXpresso`) would miss out on a growing (if not the largest) user base.

Perhaps `Tectonic` could allow the user to choose which underlying engine to use (with a warning about the slowness of LuaTeX).

**Reply by dbitouze** [reply]
> I understand, but Lua(La)TeX is not as slow as it used to be. Also, there are more and more Lua(La)TeX-specific packages, and as a result more and more users are switching to Lua(La)TeX exclusively.

And users are advised by the LaTeX team [to switch to Lua(La)TeX](https://github.com/latex3/latex2e/blob/69f76c1389ab0ff26fe885d9b2db7fce0823c15d/base/doc/ltnews40.tex#L170-L176).

**Reply by dbitouze** [reply]
> I understand, but Lua(La)TeX is not as slow as it used to be. Also, there are more and more Lua(La)TeX-specific packages, and as a result more and more users are switching to Lua(La)TeX exclusively.

Now, [`beamer`'s team advises to not use XeTeX anymore](https://github.com/josephwright/beamer/blob/dc0cae2b34d752b9c84af69dba4a95736309bafc/doc/beamerug-introduction.tex#L26) and [here](https://www.texdev.net/2024/11/05/engine-news-from-the-latex-project) is another advise from the LaTeX team to move away from XeTeX.

**Comment by lodwicz** [comment]
I've just stumbled upon tectonic (via trying to escape from TeXLive, after some time working with Pandoc), and I believe falling back to LuaTeX would truly be a fall back: XeTeX was the base for tectonic, afaics, but not necessarily forever (in a similar vein, pdfTeX was the base for LuaTeX, and LuaTeX was the base for the now actively maintained LuaMetaTeX). If we needed a scripting language, we could something else, via escape-shell. Tcl looks promising, as it is actually all strings, unlike TeX, which is all tokens.
